homeassistant/util/dt.py

```diff
@@ -309,40 +310,55 @@ def _lower_bound(arr: list[int], cmp: int) -> int | None:
     result = result.replace(hour=next_hour)
 
     if result.tzinfo in (None, UTC):
+        # Using UTC, no DST checking needed
         return result
 
-    if _datetime_ambiguous(result):
-        # This happens when we're leaving daylight saving time and local
-        # clocks are rolled back. In this case, we want to trigger
-        # on both the DST and non-DST time. So when "now" is in the DST
-        # use the DST-on time, and if not, use the DST-off time.
-        fold = 1 if now.dst() else 0
-        if result.fold != fold:
-            result = result.replace(fold=fold)
-
     if not _datetime_exists(result):
-        # This happens when we're entering daylight saving time and local
-        # clocks are rolled forward, thus there are local times that do
-        # not exist. In this case, we want to trigger on the next time
-        # that *does* exist.
-        # In the worst case, this will run through all the seconds in the
-        # time shift, but that's max 3600 operations for once per year
+        # When entering DST and clocks are turned forward.
+        # There are wall clock times that don't "exist" (an hour is skipped).
+
+        # -> trigger on the next time that 1. matches the pattern and 2. does exist
+        # for example:
+        #   on 2021.03.28 02:00:00 in CET timezone clocks are turned forward an hour
+        #   with pattern "02:30", don't run on 28 mar (such a wall time does not exist on this day)
+        #   instead run at 02:30 the next day
+
+        # We solve this edge case by just iterating one second until the result exists
+        # (max. 3600 operations, which should be fine for an edge case that happens once a year)
         return find_next_time_expression_time(
```

Comment:
This is not really *iterating* but rather *recursing* up to 3600 times, is that an issue except making debugging difficult?

Reply:
True it's actually a recursion (i wonder if cpython can perform tail call optimization here...). It's just implemented recursively due to laziness :P Otherwise it would all need to be wrapped in an ugly loop